crawler-kuing.py

```diff
@@ -0,0 +1,140 @@
+        except (KeyboardInterrupt, SystemExit):
+            print("user aborting...")
+            raise
+        except Exception as err:
+            if retry_cnt < 10:
+                retry_cnt += 1
+            else:
+                buf.close()
+                raise
+            wait_time = retry_cnt * 10.0
+            print(err)
+            print(f"[curl] sleep {wait_time} and try again...")
+            time.sleep(wait_time)
+            continue
+        break
+    res_str = buf.getvalue()
+    buf.close()
+    return res_str
+
+def extract_content(soup: BeautifulSoup) -> Dict[str, Union[str, int]]:
+    content_data = {
+        "forum": soup.title.text.split(" - ")[-2],
+        "keywords": soup.find('meta', {'name': 'keywords'})['content'] if soup.find('meta', {'name': 'keywords'}) else "",
+        "content": "\n".join([replace_dollar_tex(replace_inline_tex(replace_display_tex(msg.get_text()))) for msg in soup.select("td.t_f")])
+    }
+    return content_data
+
+def save_json(path: str, content_data: Dict[str, Union[str, int]], url: str):
+    data = {
+        "url": url,
+        "tags": content_data["keywords"].split(",") if content_data["keywords"] else [],
+        "text": content_data["content"]
+    }
+    with open(path, "w", encoding="utf-8") as f:
+        json.dump(data, f, ensure_ascii=False, sort_keys=True)
+
+def save_preview(path: str, content_data: Dict[str, Union[str, int]], url: str):
+    # put preview into HTML template
+    script_dir = os.path.dirname(__file__)
+    with open(os.path.join(script_dir, "template.html"), "r", encoding="utf-8") as f:
+        fmt_str = f.read()
+    content_html = content_data["content"].replace("\n", "</br>")
+    preview = fmt_str.replace("{PREVIEW}", content_html).replace("{URL}", url)
+    # save preview
+    with open(path, "w", encoding="utf-8") as f:
+        f.write(preview)
+
+def get_curl():
+    c = pycurl.Curl()
+    c.setopt(c.CONNECTTIMEOUT, 8)
+    c.setopt(c.TIMEOUT, 10)
+    c.setopt(c.CAINFO, certifi.where())
+    c.setopt(c.FOLLOWLOCATION, 1)
+    return c
+
+def mkdir_p(path: str):
+    try:
+        os.makedirs(path)
+    except OSError as e:
+        if e.errno == errno.EEXIST and os.path.isdir(path):
+            pass
+        else:
+            raise Exception("mkdir needs permission")
+
+def get_file_path(url: str) -> str:
+    file_id = hash(url) % DIVISIONS
+    directory = f"./tmp/{file_id}"
+    return os.path.join(directory, file_prefix) + str(file_id)
+
+def process_page(url: str, c: pycurl.Curl):
+    try:
+        page_content = curl(url, c)
+        soup = BeautifulSoup(page_content, "html.parser")
+        for pstatus in soup.select("i.pstatus"):
+            pstatus.decompose()
+        for quote in soup.select("div.quote > blockquote > font[size='2']"):
+            quote.decompose()
+        content_data = extract_content(soup)
+        if content_data["forum"] != "初等数学讨论":
+            return
+        file_path = get_file_path(url)
```

Comment:
This `file_path` is derived from hashed URL reminder, so using `file_path` as file name later may overwrite previous files. We should use an uniquely identifiable number instead (e.g., extract the thread_id rather than using URL hash).

Reply:
also, can we leave 2 newlines to separate each global functions? 

ref: https://peps.python.org/pep-0008/#blank-lines